spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -109,84 +120,138 @@ public Object postProcessAfterInitialization(final Object bean, String beanName)
 			public void doWith(final Method method) throws IllegalArgumentException, IllegalAccessException {
 				StreamListener streamListener = AnnotationUtils.findAnnotation(method, StreamListener.class);
 				if (streamListener != null) {
-					if (!method.getReturnType().equals(Void.TYPE)) {
-						Assert.isTrue(method.getAnnotation(Input.class) == null,
-								"A @StreamListener may never be annotated with @Input." +
-										"If it should listen to a specific input, use the value of @StreamListener " +
-										"instead.");
-					}
+					Assert.isTrue(method.getAnnotation(Input.class) == null, INPUT_AT_STREAM_LISTENER);
+					String methodAnnotatedInboundName = getInboundElementNameFromMethod(streamListener);
+					String methodAnnotatedOutboundName = getOutboundElementNameFromMethod(method);
+					int inputAnnotationCount = StreamListenerMethodUtils.inputAnnotationCount(method);
+					int outputAnnotationCount = StreamListenerMethodUtils.outputAnnotationCount(method);
+					boolean isDeclarative = isDeclarativeStreamListenerMethod(method, methodAnnotatedInboundName, methodAnnotatedOutboundName);
+					StreamListenerMethodUtils.assertStreamListenerMethod(method, inputAnnotationCount, outputAnnotationCount, methodAnnotatedInboundName, methodAnnotatedOutboundName, isDeclarative);
 					Class<?>[] parameterTypes = method.getParameterTypes();
-					if (StringUtils.hasText(streamListener.value())) {
-						for (int i = 0; i < parameterTypes.length; i++) {
-							MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, i);
-							Assert.isTrue(methodParameter.getParameterAnnotation(Input.class) == null &&
-											methodParameter.getParameterAnnotation(Output.class) == null,
-									"A message handling @StreamListener method cannot have parameters annotated " +
-											"with @Input or @Output");
-						}
-						if (!method.getReturnType().equals(Void.TYPE)) {
-							Assert.isTrue(method.getAnnotation(Output.class) == null,
-									"A message handling @StreamListener method cannot be annotated with @Output");
+					if (!method.getReturnType().equals(Void.TYPE)) {
+						if (!StringUtils.hasText(methodAnnotatedOutboundName)) {
+							if (outputAnnotationCount == 0) {
+								throw new IllegalArgumentException(RETURN_TYPE_NO_OUTBOUND_SPECIFIED);
+							}
+							Assert.isTrue((outputAnnotationCount == 1), RETURN_TYPE_MULTIPLE_OUTBOUND_SPECIFIED);
 						}
-						registerHandlerMethodOnListenedChannel(method, streamListener, bean);
+					}
+					if (isDeclarative && (!StringUtils.hasText(methodAnnotatedInboundName) || parameterTypes.length == 1)) {
```

Comment:
Actually, at a second glance, this seems wrong - if a method would normally be declarative (i.e. has the right arguments/annotations) but has either more than one argument or declares an input via annotation (e.g. `@StreamListener("input"))` we set up the handler - we should never set up a handler for a declarative method.